hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/TransactionUtils.java

```diff
@@ -96,11 +97,122 @@ public static Option<HoodieCommitMetadata> resolveWriteConflictIfAny(
       });
       LOG.info("Successfully resolved conflicts, if any");
 
+      // Resolve schema.
+      Schema schemaOfCommitMetadata = resolveConcurrentSchemaEvolution(
+          table, config, lastCompletedTxnOwnerInstant, new TableSchemaResolver(table.getMetaClient()));
+      if (thisCommitMetadata.isPresent()) {
+        thisCommitMetadata.get().addMetadata(HoodieCommitMetadata.SCHEMA_KEY, schemaOfCommitMetadata.toString());
+      }
+      thisOperation.getCommitMetadataOption().get().addMetadata(
+          HoodieCommitMetadata.SCHEMA_KEY, schemaOfCommitMetadata.toString());
       return thisOperation.getCommitMetadataOption();
     }
     return thisCommitMetadata;
   }
 
+  /**
+   * Resolve concurrent schema evolution. If it is resolvable, return the schema to be set in the commit metadata.
+   * Otherwise, throw a {@link HoodieWriteConflictException}.
+   *
+   * @param table The Hoodie table.
+   * @param config The Hoodie write configuration.
+   * @param lastCompletedTxnOwnerInstant The last completed transaction owner instant.
+   * @param schemaResolver The table schema resolver.
+   * @throws HoodieWriteConflictException If there is a concurrent schema evolution.
+   */
+  static Schema resolveConcurrentSchemaEvolution(
+      HoodieTable table,
+      HoodieWriteConfig config,
+      Option<HoodieInstant> lastCompletedTxnOwnerInstant,
+      TableSchemaResolver schemaResolver) {
+    HoodieInstant lastCompletedInstantsAtTxnStart = lastCompletedTxnOwnerInstant.isPresent() ? lastCompletedTxnOwnerInstant.get() : null;
+    HoodieInstant lastCompletedInstantsAtTxnValidation = table.getMetaClient().reloadActiveTimeline().getCommitsTimeline().filterCompletedInstants()
```

Comment:
Could we reuse `ConcurrentOperation` instances for both schema and data conflict detection?